lib/bloc/backup/backup_bloc.dart

```diff
@@ -325,21 +369,227 @@ class BackupBloc {
+        exception = SignInCancelledException();
+      } else if (e.code == _signInFailedMessage ||
+          e.message == _signInFailedCode) {
+        exception = SignInFailedException(
+          _backupSettingsController.value.backupProvider,
+        );
+      } else if (e.code == _empty || exception == _empty) {
+        exception = NoBackupFoundException();
+      }
+      action.resolveError(exception);
+    } catch (error) {
+      log.warning("Failed to list snapshots.", error);
+      action.resolveError("Failed to list snapshots.");
+    }
+  }
+
+  Future _restoreBackup(RestoreBackup action) async {
+    try {
+      final snapshot = action.restoreRequest.snapshot;
+      final key = action.restoreRequest.encryptionKey.key;
+      log.info('snapshotInfo with timestamp: ${snapshot.modifiedTime}');
+      if (key != null) log.info('using key with length: ${key.length}');
+      _setBackupKeyType(snapshot);
+
+      _clearAppData();
+
+      await _breezLib.restore(snapshot.nodeID, key);
+      await _restoreAppData();
+
+      _updateLastBackupTime(snapshot.modifiedTime);
+
+      action.resolve(true);
+    } on FileSystemException catch (error) {
+      if (error.message.contains("Failed to decode data using encoding")) {
+        log.warning(
+            "Failed to restore backup. Incorrect mnemonic phrase.", error);
+        _clearAppData();
+        await _resetBackupKey();
+        action.resolveError(
+          getSystemAppLocalizations().enter_backup_phrase_error,
+        );
+      }
+    } catch (error) {
+      log.warning("Failed to restore backup.", error);
+      _clearAppData();
+      await _resetBackupKey();
+      action.resolveError("Failed to restore backup.");
+    }
+  }
+
+  void _setBackupKeyType(SnapshotInfo snapshot) {
+    BackupSettings backupSettings = _backupSettingsController.value;
+    if (snapshot.encrypted && snapshot.encryptionType.startsWith("Mnemonics")) {
+      backupSettings = backupSettings.copyWith(keyType: BackupKeyType.PHRASE);
+    } else if (backupSettings.backupKeyType != BackupKeyType.NONE) {
+      /// This also sets keyType of backups encrypted with PIN to
+      /// BackupKeyType.NONE as they are are non-secure & deprecated
+      backupSettings = backupSettings.copyWith(keyType: BackupKeyType.NONE);
+    }
+    _backupSettingsController.add(backupSettings);
+  }
+
+  void _updateLastBackupTime(String modifiedTime) {
+    _backupStateController.add(
+      _backupStateController.value?.copyWith(
+        lastBackupTime: DateTime.tryParse(modifiedTime).toLocal(),
+      ),
+    );
+  }
+
+  Future _signOut(SignOut action) async {
+    log.info("Signing out of Google Drive");
+    await _breezLib.signOut().catchError(
+      (error) {
+        log.warning("Failed to sign out.", error);
+        _promptBackupController.add(action.promptOnError);
+      },
+    );
+
+    backupServiceNeedLoginSink.add(true);
+    action.resolve(null);
+  }
+
+  Future _signIn() async {
```

Comment:
TODO: force signIn when backupServiceNeedLogin is true, however, this requires changes on iOS as force sign in isn't implemented there yet*. Afterwards, use this function on _listSnapshots. Remove all sign in error handling from _listSnapshots & _backupNow.

*: breezLib.signOut() can be called manually if sign in is needed